ovn-tester/cms/openstack/openstack.py

```diff
@@ -0,0 +1,555 @@
+import logging
+import uuid
+from dataclasses import dataclass
+from typing import List, Optional, Dict
+
+import netaddr
+
+from randmac import RandMac
+
+from ovn_sandbox import PhysicalNode
+from ovn_utils import (
+    DualStackSubnet,
+    LSwitch,
+    LSPort,
+    PortGroup,
+    LRouter,
+    LRPort,
+    DualStackIP,
+)
+from ovn_workload import ChassisNode, Cluster
+
+log = logging.getLogger(__name__)
+
+OVN_HEATER_CMS_PLUGIN = 'OpenStackCloud'
+
+
+@dataclass
+class NeutronNetwork:
+    """Group of OVN objects that logically belong to same Neutron network."""
+
+    network: LSwitch
+    ports: Dict[str, LSPort]
+    security_group: Optional[PortGroup] = None
+
+    def __post_init__(self):
+        self._ip_pool = self.network.cidr.iter_hosts()
+        self.gateway = self.next_host_ip()
+
+    def next_host_ip(self) -> netaddr.IPAddress:
+        """Return next available host IP in this network's range."""
+        return next(self._ip_pool)
+
+
+class Project:
+    """Represent network components of an OpenStack Project aka. Tenant."""
+
+    def __init__(
+        self,
+        networks: List[NeutronNetwork] = None,
+        router: Optional[LRouter] = None,
+    ):
+        self.networks: List[NeutronNetwork] = (
+            [] if networks is None else networks
+        )
+        self.router = router
+        self._id = str(uuid.uuid4())
+
+    @property
+    def uuid(self) -> str:
+        """Return arbitrary UUID assigned to this Openstack Project."""
+        return self._id
+
+
+class OpenStackCloud(Cluster):
+    """Representation of Openstack cloud/deployment."""
+
+    def __init__(self, cluster_cfg, central, brex_cfg, az):
+        super().__init__(cluster_cfg, central, brex_cfg, az)
+        self.router = None
+        self.external_port: Optional[LRPort] = None
+        self._projects: List[Project] = []
+
+        self._int_net_base = DualStackSubnet(cluster_cfg.node_net)
+        self._int_net_offset = 0
+
+        self._ext_net_pool_index = 0
+
+    def add_cluster_worker_nodes(self, workers: List[PhysicalNode]):
+        """Add OpenStack flavored worker nodes as per configuration."""
+        # Allocate worker IPs after central and relay IPs.
+        mgmt_ip = (
+            self.cluster_cfg.node_net.ip
+            + 2
+            + len(self.central_nodes)
+            + len(self.relay_nodes)
+        )
+
+        protocol = "ssl" if self.cluster_cfg.enable_ssl else "tcp"
+        # XXX: To facilitate network nodes we'd have to make bigger change
```

Comment:
We had some discussions internally about potentially removing the ovn-fake-multinode dependency and start OVN components directly from ovn-heater.  Whenever (if) we do that it might be a good moment for adding support for network nodes too.